editor/src/components/canvas/remix/remix-utils.tsx

```diff
@@ -99,3 +128,397 @@ export function getRoutesFromRouteManifest(
+  }
+}
+
+export interface RouteModuleCreator {
+  filePath: string
+  executionScopeCreator: (projectContents: ProjectContentTreeRoot) => ExecutionScope
+}
+
+export interface RouteIdsToModuleCreators {
+  [routeId: string]: RouteModuleCreator
+}
+
+export interface RouteModulesWithRelativePaths {
+  [routeId: string]: {
+    relativePath: ElementPath
+    filePath: string
+  }
+}
+
+export interface GetRoutesAndModulesFromManifestResult {
+  routeModuleCreators: RouteIdsToModuleCreators
+  routes: Array<DataRouteObject>
+  routeModulesToRelativePaths: RouteModulesWithRelativePaths
+  routingTable: RemixStaticRoutingTable
+}
+
+export function getDefaultExportNameAndUidFromFile(
+  projectContents: ProjectContentTreeRoot,
+  filePath: string,
+): { name: string; uid: string } | null {
+  const file = getProjectFileByFilePath(projectContents, filePath)
+  if (file == null || file.type != 'TEXT_FILE' || file.lastParseSuccess == null) {
+    return null
+  }
+
+  const defaultExportName =
+    file.lastParseSuccess.exportsDetail.find(
+      (e): e is ExportDefaultFunctionOrClass => e.type === 'EXPORT_DEFAULT_FUNCTION_OR_CLASS',
+    )?.name ?? null
+
+  if (defaultExportName == null) {
+    return null
+  }
+
+  const elementUid = file.lastParseSuccess.topLevelElements.find(
+    (t): t is UtopiaJSXComponent =>
+      t.type === 'UTOPIA_JSX_COMPONENT' && t.name === defaultExportName,
+  )?.rootElement.uid
+  if (elementUid == null) {
+    return null
+  }
+
+  return { name: defaultExportName, uid: elementUid }
+}
+
+export function getDefaultExportedTopLevelElement(file: TextFile): JSXElementChild | null {
+  if (file.fileContents.parsed.type !== 'PARSE_SUCCESS') {
+    return null
+  }
+
+  const defaultExportName =
+    file.fileContents.parsed.exportsDetail.find(
+      (e): e is ExportDefaultFunctionOrClass => e.type === 'EXPORT_DEFAULT_FUNCTION_OR_CLASS',
+    )?.name ?? null
+
+  if (defaultExportName == null) {
+    return null
+  }
+
+  return (
+    file.fileContents.parsed.topLevelElements.find(
+      (t): t is UtopiaJSXComponent =>
+        t.type === 'UTOPIA_JSX_COMPONENT' && t.name === defaultExportName,
+    )?.rootElement ?? null
+  )
+}
+
+function addLoaderAndActionToRoute(
+  routes: DataRouteObject[],
+  routeId: string,
+  loader: LoaderFunction | undefined,
+  action: ActionFunction | undefined,
+) {
+  routes.forEach((route) => {
+    if (route.id === routeId) {
+      route.action = action
+      route.loader = loader
+    } else {
+      addLoaderAndActionToRoute(route.children ?? [], routeId, loader, action)
+    }
+  })
+}
+
+function findOutletAmongJSXElementChildren(
+  parentUid: string,
+  isOutlet: (e: JSXElementChild) => boolean,
+  children: JSXElementChild[],
+): ElementPathPart | null {
+  for (const child of children) {
+    const path = findPathToOutlet(isOutlet, child)
+    if (path != null) {
+      return [parentUid, ...path]
+    }
+  }
+  return null
+}
+
+export function isRemixOutletElement(
```

Comment:
I believe we agreed that this should live in either `element-metadata-utils.ts` or `project-file-utils.ts`

Reply:
Do we really want to add library specific stuff to those?